cirq/google/calibration/phased_fsim.py

```diff
@@ -76,3 +85,102 @@ def override_by(self, other: 'PhasedFSimParameters') -> 'PhasedFSimParameters':
         """
         return other.merge_with(self)
+
+
+@json_serializable_dataclass(frozen=True)
+class FloquetPhasedFSimCalibrationOptions:
+    characterize_theta: bool
+    characterize_zeta: bool
+    characterize_chi: bool
+    characterize_gamma: bool
+    characterize_phi: bool
+
+
+# TODO: Fix json serialization (the default one doesn't work with tuples as dictionary keys).
+# TODO: Add start and end calibration timestamp
+# TODO: Add export to Panda's data frame
+@json_serializable_dataclass(frozen=True)
+class PhasedFSimCalibrationResult:
+    # TODO: Instead of parameters and gate, should that just be cirq.Moment validated at construction time? (This is
+    #  slightly problematic with frozen data classes).
+    parameters: Dict[Tuple[Qid, Qid], PhasedFSimParameters]
+    gate: Gate
+    gate_set: SerializableGateSet
+
+    def get_parameters(self, a: Qid, b: Qid) -> Optional['PhasedFSimParameters']:
+        if (a, b) in self.parameters:
+            return self.parameters[(a, b)]
+        elif (b, a) in self.parameters:
+            return self.parameters[(b, a)].parameters_for_qubits_swapped()
+        else:
+            return None
+
+
+@json_serializable_dataclass(frozen=True)
+class PhasedFSimCalibrationRequest(abc.ABC):
+    gate: Gate  # Any gate which can be described by cirq.PhasedFSim
+    gate_set: SerializableGateSet
+    pairs: Tuple[Tuple[Qid, Qid]]
+
+    @abc.abstractmethod
+    def to_calibration_layer(self, handler_name: str) -> CalibrationLayer:
+        pass
+
+    @abc.abstractmethod
+    def parse_result(self, result: CalibrationResult) -> PhasedFSimCalibrationResult:
+        pass
+
+
+@json_serializable_dataclass(frozen=True)
+class FloquetPhasedFSimCalibrationResult(PhasedFSimCalibrationResult):
+    options: FloquetPhasedFSimCalibrationOptions
+
+
+@json_serializable_dataclass(frozen=True)
+class FloquetPhasedFSimCalibrationRequest(PhasedFSimCalibrationRequest):
+    options: FloquetPhasedFSimCalibrationOptions
+
+    def to_calibration_layer(self, handler_name: str) -> CalibrationLayer:
+        circuit = Circuit([self.gate.on(*pair) for pair in self.pairs])
+        return CalibrationLayer(
+            calibration_type='floquet_phased_fsim_characterization',
+            program=circuit,
+            args={
+                'est_theta': self.options.characterize_theta,
+                'est_zeta': self.options.characterize_zeta,
+                'est_chi': self.options.characterize_chi,
+                'est_gamma': self.options.characterize_gamma,
+                'est_phi': self.options.characterize_phi,
+                'readout_corrections': True
+            }
+        )
+
+    def parse_result(self, result: CalibrationResult) -> PhasedFSimCalibrationResult:
+        decoded = collections.defaultdict(lambda: {})
+        for keys, values in result.metrics['angles'].items():
+            for key, value in zip(keys, values):
+                match = re.match(r'(\d+)_(.+)', key)
```

Comment:
We probably should document the key string format somewhere.

Reply:
I intend to change it, I would prefer to wait with that until it's stable.